src/server/launchpad/client.js

```diff
@@ -0,0 +1,347 @@
+  const base_parsed = url.parse(base_uri);
+  const parsed = url.parse(uri);
+
+  if (parsed.host !== null) {
+    // e.g. 'http://www.example.com/api/devel/foo';
+    // Don't try to insert the service base into what was an absolute URL.
+    // So 'http://www.example.com/foo' remains unchanged.
+  } else {
+    if (!parsed.pathname.startsWith('/')) {
+      // e.g. 'api/devel/foo' or 'foo'
+      parsed.pathname = '/' + parsed.pathname;
+    }
+    if (!parsed.pathname.startsWith(service_base)) {
+      // e.g. '/foo'
+      parsed.pathname = service_base + parsed.pathname;
+    }
+  }
+
+  parsed.protocol = base_parsed.protocol;
+  parsed.host = base_parsed.host;
+
+  return url.format(parsed);
+}
+
+class ResourceError extends Error {
+  constructor(response, client, uri, method) {
+    super(`${uri} returned HTTP status ${response.status}`);
+    this.response = response;
+    this.client = client;
+    this.uri = uri;
+    this.method = method;
+  }
+}
+
+export function wrapResourceOnSuccess(response, client, uri, method) {
+  const media_type = response.headers.get('Content-Type');
+  if (media_type.startsWith('application/json')) {
+    return response.json().then(representation => {
+      // If the object fetched has a self_link, make that the object's uri
+      // for use in other api methods off of that object.  During a PATCH
+      // request the caller is the object.  Leave the original_uri alone.
+      // Otherwise make the uri the object coming back.
+      if (representation.self_link && method !== 'PATCH') {
+        uri = representation.self_link;
+      }
+      return client.wrap_resource(uri, representation);
+    });
+  } else {
+    return response.text();
+  }
+}
+
+// The resources that come together to make Launchpad.
+
+/** The base class for objects retrieved from Launchpad's web service. */
+export class Resource {
+  /** Initialize a resource with its representation and URI. */
+  init(client, representation, uri) {
+    this.lp_client = client;
+    this.uri = uri;
+    for (const key of Object.keys(representation)) {
+      this[key] = representation[key];
+    }
+  }
+
+  /** Get the result of a named GET operation on this resource. */
+  named_get(operation_name, config) {
+    return this.lp_client.named_get(this.uri, operation_name, config);
+  }
+
+  /** Trigger a named POST operation on this resource. */
+  named_post(operation_name, config) {
+    return this.lp_client.named_post(this.uri, operation_name, config);
+  }
+}
+
+/** The root of the Launchpad web service. */
+export class Root extends Resource {
+  constructor(client, representation, uri) {
+    super();
+    this.init(client, representation, uri);
+  }
+}
+
+/** A grouped collection of objects from the Launchpad web service. */
+export class Collection extends Resource {
+  constructor(client, representation, uri) {
+    super();
+    this.init(client, representation, uri);
+    for (const i in this.entries.keys()) {
```

Comment:
Any reason for iterating over `keys` instead of simply doing good'ol `for (const i = 0; i < this.entries.length)` or use `this.entries.forEach()`?